k2/csrc/hash.h

```diff
@@ -1014,6 +1014,349 @@ class Hash {
+      once (necessary because there is no concept of a "tombstone".
+
+   Some notes on usage:
+
+   You use it by: constructing it, obtaining its Accessor with GetAccessor();
+   and inside kernels (or host code), calling functions Insert(), Find() or
+   Delete() of the Accessor object.  Resizing is not automatic; it is the
+   user's responsibility to make sure the hash does not get too full
+   (which could cause assertion failures in kernels, and will be very slow).
+
+   Some implementation notes:
+    - When accessing hash[key], we use bucket_index == key % num_buckets,
+      bucket_inc = 1 | (((key * 2) / num_buckets) ^ key).
+    - If the bucket at `bucket_index` is occupied, we look in locations
+      `(bucket_index + n * bucket_inc)%num_buckets` for n = 1, 2, ...;
+      this choice ensures that if multiple keys hash to the same bucket,
+      they don't all access the same sequence of locations; and bucket_inc
+      being odd ensures we eventually try all locations (of course for
+      reasonable hash occupancy levels, we shouldn't ever have to try
+      more than two or three).
+
+*/
+class Hash64 {
+ public:
+  /* Constructor.  Context can be for CPU or GPU.
+
+     @param [in] num_buckets   Number of buckets in the hash; must be
+                a power of 2 and >= 128 (this limit was arbitrarily chosen).
+                The number of items in the hash cannot exceed the number of
+                buckets, or the code will loop infinitely when you try to add
+                items; aim for less than 50% occupancy.
+  */
+  Hash64(ContextPtr c, int32_t num_buckets) {
+    K2_CHECK_GE(num_buckets, 128);
+    data_ = Array1<uint64_t>(c, num_buckets * 2, ~(uint64_t)0);
+    int32_t n = 2;
+    for (buckets_num_bitsm1_ = 0; n < num_buckets;
+         n *= 2, buckets_num_bitsm1_++) {
+    }
+    K2_CHECK_EQ(num_buckets, 2 << buckets_num_bitsm1_)
+        << " num_buckets must be a power of 2.";
+  }
+
+  // Only to be used prior to assignment.
+  Hash64() = default;
+
+  int32_t NumBuckets() const { return data_.Dim() / 2; }
+
+  // Returns data pointer; for testing..
+  uint64_t *Data() { return data_.Data(); }
+
+  // Shallow copy
+  Hash64 &operator=(const Hash64 &src) = default;
+  // Copy constructor (shallow copy)
+  explicit Hash64(const Hash64 &src) = default;
+
+  ContextPtr &Context() const { return data_.Context(); }
+
+  class Accessor {
+   public:
+    Accessor(Hash64 &hash)
+        : data_(hash.data_.Data()),
+          num_buckets_mask_(uint32_t(hash.NumBuckets()) - 1),
+          buckets_num_bitsm1_(hash.buckets_num_bitsm1_) {}
+
+    // Copy constructor
+    Accessor(const Accessor &src) = default;
+
+    /*
+     Try to insert pair (key,value) into hash.
+       @param [in] key  Key into hash, it is an error if ~key == 0, i.e. if all
+                        the allowed bits of `key` are set.
+       @param [in] value  Value to set, it is an error if ~value == 0, i.e. if
```

Comment:
I think this "it is an error" stuff on "value" may not be necessary and the stipulation on 'key' only may be enough??
Is there a particular race condition that it solves?

Reply:
`~(unint64_t)0` is used as `invalid`,  in `Find` it will wait in a loop until reading a value different from `~(unint64)0`.

https://github.com/k2-fsa/k2/blob/bb11575cf4fede0da96d877893a66347db3faacf/k2/csrc/hash.h#L1179-L1195